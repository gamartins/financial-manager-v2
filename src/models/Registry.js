export default class Registry {
  static TYPE = {
    PROVENTO: 0,
    DESCONTO: 1,
  }

  static STATUS = {
    NOT_PAID: 0,
    PAID: 1,
    PAID_FINANCIAL_CASH: 2,
    RESERVERD_SAVINGS: 3,
  }

  static getStatusString(status) {
    switch (status) {
    case Registry.STATUS.NOT_PAID:
      return 'NÃ£o pago'
    case Registry.STATUS.PAID:
      return 'Pago'
    case Registry.STATUS.PAID_FINANCIAL_CASH:
      return 'Depositar em poup'
    case Registry.STATUS.RESERVERD_SAVINGS:
      return 'Depositado em poup'
    default:
      return ''
    }
  }
}
