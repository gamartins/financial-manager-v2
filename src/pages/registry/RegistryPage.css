.Registry .actions {
    display: flex;
    justify-content: space-between;
}

.Registry .actions .btn {
    margin-bottom: 1rem;
}