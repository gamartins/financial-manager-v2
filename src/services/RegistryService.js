const key = 'fmv2:registry'

const get = dateQuery => {
  const dataList = JSON.parse(localStorage.getItem(`${key}${dateQuery}`))

  if (!dataList) {
    localStorage.setItem(`${key}${dateQuery}`, JSON.stringify([]))
    return []
  }

  return dataList
}

const save = (dataList, dateQuery) => {
  localStorage.setItem(`${key}${dateQuery}`, JSON.stringify(dataList))
}

export default {
  get,
  save,
}
