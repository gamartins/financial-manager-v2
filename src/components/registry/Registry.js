import React, { Component } from 'react'

import addMonths from 'date-fns/addMonths'

import RegistryForm from './RegistryForm'
import RegistryList from './RegistryList'
import DatePagination from '../date-pagination/DatePagination'
import RegistryService from '../../services/RegistryService'

export default class Registry extends Component {

  static TYPE = {
    PROVENTO: 0,
    DESCONTO: 1,
  }

  constructor(props) {
    super(props)

    this.state = {
      date: new Date(),
      dataList: [],
      formValues: {
        id: null,
        name: '',
        price: 0,
        type: Registry.TYPE.DESCONTO,
        category: ''
      }
    }
  }

  componentDidMount() {
    this.getAll()
  }

  getAll() {
    const dateQuery = `${this.state.date.getFullYear()}${this.state.date.getMonth()}`

    const dataList = RegistryService.get(dateQuery)
    
    this.setState({ dataList })
  }

  getNext() {
    this.setState(
      (state) => ({ date: addMonths(state.date, 1) }),
      () => this.getAll())
  }

  getPrevious() {
    this.setState(
      (state) => ({ date: addMonths(state.date, -1) }),
      () => this.getAll())
  }

  setFormValue(id) {
    const index = this.state.dataList.findIndex(item => item.id === id)

    if (index === -1) {
      return
    }

    const newFormValues = { ...this.state.dataList[index] }
    
    this.setState({ formValues: newFormValues })
  }

  resetFormValue() {
    this.setState({
      formValues: {
        id: null,
        name: '',
        price: 0,
        type: Registry.TYPE.DESCONTO,
        category: '',
      }
    })
  }

  save(values) {
    return values.id ? this.edit(values) : this.add(values)
  }

  add(values) {
    const dateQuery = `${this.state.date.getFullYear()}${this.state.date.getMonth()}`
    const dataList = [ ...this.state.dataList ]
    dataList.push({
      id: Math.ceil((Math.random() * 100000)),
      name: values.name,
      price: values.price,
      type: parseInt(values.type),
      category: parseInt(values.category),
    })

    RegistryService.save(dataList, dateQuery)
    this.setState({ dataList: dataList })
    this.resetFormValue()

    return Promise.resolve()
  }

  edit(values) {
    const dateQuery = `${this.state.date.getFullYear()}${this.state.date.getMonth()}`
    const index = this.state.dataList.findIndex(item => item.id === values.id)
    const dataList = [ ...this.state.dataList ]
    dataList[index] = {
      id: values.id,
      name: values.name,
      price: values.price,
      type: parseInt(values.type),
      category: parseInt(values.category),
    }

    RegistryService.save(dataList, dateQuery)
    this.setState({ dataList: dataList })
    this.resetFormValue()

    return Promise.resolve()
  }

  remove(id) {
    const index = this.state.dataList.findIndex(item => item.id === id)

    if (index !== -1) {
      this.state.dataList.splice(index, 1)
      this.setState({ dataList: this.state.dataList })
      this.resetFormValue()
    }    
  }

  render() {
    return (
      <div>
        <RegistryForm
          initialValues={this.state.formValues}
          saveItem={values => this.save(values)} />
        <DatePagination
          date={this.state.date}
          format={'MMM'}
          previous={() => this.getPrevious()}
          next={() => this.getNext()}/>
        <RegistryList
          dataList={this.state.dataList}
          removeItem={id => this.remove(id)}
          editItem={item => this.setFormValue(item)} />
      </div>
    )
  }
}
