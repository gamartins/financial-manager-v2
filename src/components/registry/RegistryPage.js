import React, { Component } from 'react'

import addMonths from 'date-fns/addMonths'
import { Button } from 'react-bootstrap'

import RegistryForm from './RegistryForm'
import RegistryList from './RegistryList'
import DatePagination from '../date-pagination/DatePagination'
import RegistryService from '../../services/RegistryService'
import Registry from '../../models/Registry'

import './RegistryPage.css'

export default class RegistryPage extends Component {
  constructor(props) {
    super(props)

    this.state = {
      date: new Date(),
      dataList: [],
      showModal: false,
      formValues: {
        id: null,
        name: '',
        price: 0,
        type: Registry.TYPE.DESCONTO,
        category: '',
        status: Registry.STATUS.NOT_PAID,
      },
    }
  }

  componentDidMount() {
    this.getAll()
  }

  getAll() {
    const dateQuery = `${this.state.date.getFullYear()}${this.state.date.getMonth()}`

    const dataList = RegistryService.get(dateQuery) || []

    this.setState({ dataList })
  }

  getNext() {
    this.setState(
      state => ({ date: addMonths(state.date, 1) }),
      () => this.getAll(),
    )
  }

  getPrevious() {
    this.setState(
      state => ({ date: addMonths(state.date, -1) }),
      () => this.getAll(),
    )
  }

  setFormValue(id) {
    const index = this.state.dataList.findIndex(item => item.id === id)

    if (index === -1) {
      return
    }

    this.setState(prevState => {
      this.toggleModal()

      const newFormValues = prevState.dataList[index]

      return { formValues: newFormValues }
    })
  }

  resetFormValue() {
    this.setState({
      formValues: {
        id: null,
        name: '',
        price: 0,
        type: Registry.TYPE.DESCONTO,
        category: '',
        status: Registry.STATUS.NOT_PAID,
      },
    })
  }

  save(values) {
    return values.id ? this.edit(values) : this.add(values)
  }

  add(values) {
    const dateQuery = `${this.state.date.getFullYear()}${this.state.date.getMonth()}`
    const item = {
      id: Math.ceil((Math.random() * 100000)),
      name: values.name,
      price: values.price,
      type: parseInt(values.type),
      category: parseInt(values.category),
      status: parseInt(values.status),
    }

    this.setState(
      prevState => ({ dataList: [...prevState, item] }),
      () => RegistryService.save(this.state.dataList, dateQuery),
    )

    this.resetFormValue()
    this.toggleModal()

    return Promise.resolve()
  }

  edit(values) {
    const dateQuery = `${this.state.date.getFullYear()}${this.state.date.getMonth()}`
    const index = this.state.dataList.findIndex(item => item.id === values.id)
    const item = {
      id: values.id,
      name: values.name,
      price: values.price,
      type: parseInt(values.type),
      category: parseInt(values.category),
      status: parseInt(values.status),
    }

    this.setState(
      prevState => {
        const modifiedList = prevState
        modifiedList[index] = item
        return { dataList: modifiedList }
      },
      () => RegistryService.save(this.state.dataList, dateQuery),
    )

    this.resetFormValue()
    this.toggleModal()

    return Promise.resolve()
  }

  remove(id) {
    const index = this.state.dataList.findIndex(item => item.id === id)

    if (index !== -1) {
      this.setState(prevState => ({ dataList: prevState.dataList.splice(index, 1) }))
      this.resetFormValue()
    }
  }

  toggleModal() {
    this.setState(
      state => ({ showModal: !state.showModal }),
      () => {
        if (!this.state.showModal) {
          this.resetFormValue()
        }
      },
    )
  }

  render() {
    return (
      <div className="Registry">
        <RegistryForm
          showModal={this.state.showModal}
          toggleModal={() => this.toggleModal()}
          initialValues={this.state.formValues}
          saveItem={values => this.save(values)}
        />

        <div className="actions">
          <Button id="actions-add" onClick={() => this.toggleModal()}>
            Novo
          </Button>

          <DatePagination
            date={this.state.date}
            format="MMM"
            previous={() => this.getPrevious()}
            next={() => this.getNext()}
          />
        </div>

        <RegistryList
          dataList={this.state.dataList}
          removeItem={id => this.remove(id)}
          editItem={item => this.setFormValue(item)}
        />
      </div>
    )
  }
}
