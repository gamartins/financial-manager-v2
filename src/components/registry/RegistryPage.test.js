import React from 'react'
import { act } from 'react-dom/test-utils'
import Enzyme, { mount, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import RegistryPage from './RegistryPage'
import RegistryService from '../../services/RegistryService'
import Registry from '../../models/Registry'
import registerIcons from '../../services/FontAwesome'

jest.mock('../../services/RegistryService')

registerIcons()

Enzyme.configure({ adapter: new Adapter() })

RegistryService.get.mockImplementation(() => [
  {
    id: 1,
    name: 'Registry 01',
    price: 2000,
    type: Registry.TYPE.PROVENTO,
    category: 1,
  },
  {
    id: 2,
    name: 'Registry 02',
    price: 69,
    type: Registry.TYPE.DESCONTO,
    category: 1,
  },
])

describe('<RegistryPage />', () => {
  test('should be created', () => {
    const wrapper = mount(<RegistryPage />)
    expect(wrapper).toMatchSnapshot()
  })

  test('should get registry list on componentDidMount', () => {
    const wrapper = shallow(<RegistryPage />)
    expect(wrapper.instance().state.dataList).toHaveLength(2)
  })

  test('should find and set form value with item selected', () => {
    const wrapper = shallow(<RegistryPage />)
    wrapper.instance().setFormValue(2)

    expect(wrapper.instance().state.formValues).toEqual({
      id: 2,
      name: 'Registry 02',
      price: 69,
      type: Registry.TYPE.DESCONTO,
      category: 1,
    })
  })

  test('should call add when registry dont exists', () => {
    const addMock = jest.fn()
    const wrapper = shallow(<RegistryPage />)
    wrapper.instance().add = addMock

    wrapper.instance().save({
      id: null,
      name: 'Registry 02',
      price: 69,
      type: Registry.TYPE.DESCONTO,
      category: 1,
    })

    expect(addMock).toHaveBeenCalled()
  })

  test('should call edit when registry already exists', () => {
    const editMock = jest.fn()
    const wrapper = shallow(<RegistryPage />)
    wrapper.instance().edit = editMock

    wrapper.instance().save({
      id: 2,
      name: 'Registry 02',
      price: 69,
      type: Registry.TYPE.DESCONTO,
      category: 1,
    })

    expect(editMock).toHaveBeenCalled()
  })

  test('should remove one element when call remove', () => {
    const wrapper = mount(<RegistryPage />)

    expect(wrapper.instance().state.dataList).toHaveLength(2)

    wrapper.instance().remove(2)

    expect(wrapper.instance().state.dataList).toHaveLength(1)
  })

  test('should call getAll with the next month query after getNext event', () => {
    const wrapper = shallow(<RegistryPage />)
    wrapper.setState({ date: new Date(2019, 11, 18) })

    const mockCallback = jest.fn(() => {})
    RegistryService.get = mockCallback

    wrapper.instance().getNext()

    expect(mockCallback).toHaveBeenCalledWith('20200')
  })

  test('should call getAll with the next month query after getNext event', () => {
    const wrapper = shallow(<RegistryPage />)
    wrapper.setState({ date: new Date(2019, 11, 18) })

    const mockCallback = jest.fn(() => {})
    RegistryService.get = mockCallback

    wrapper.instance().getPrevious()

    expect(mockCallback).toHaveBeenCalledWith('201910')
  })

  test('should show <RegistryForm /> modal when click on new button', async () => {
    const wrapper = shallow(<RegistryPage />)
    expect(wrapper.instance().state.showModal).toBeFalsy()

    await act(async () => {
      wrapper.find('.actions #actions-add').simulate('click', { preventDefault: () => {} })
    })

    expect(wrapper.instance().state.showModal).toBeTruthy()
  })

  test('should reset form when hide modal', () => {
    const wrapper = shallow(<RegistryPage />)
    wrapper.setState({
      dataList: [],
      showModal: true,
      formValues: {
        id: null,
        name: 'Some item',
        price: 59.90,
        type: Registry.TYPE.DESCONTO,
        category: 1,
        status: Registry.STATUS.NOT_PAID,
      },
    })

    wrapper.instance().toggleModal()

    expect(wrapper.instance().state.formValues).toEqual({
      id: null,
      name: '',
      price: 0,
      type: Registry.TYPE.DESCONTO,
      category: '',
      status: Registry.STATUS.NOT_PAID,
    })
  })
})
