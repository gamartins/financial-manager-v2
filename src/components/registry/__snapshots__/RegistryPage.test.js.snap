// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegistryPage /> should be created 1`] = `ReactWrapper {}`;
